File: desktop/native/SurveyController.App/ViewModels/OptionWeight.cpp
#include "pch.h"
#include "OptionWeight.h"

#if __has_include("OptionWeight.g.cpp")
#include "OptionWeight.g.cpp"
#endif

namespace winrt::SurveyController::App::implementation
{
    OptionWeight::OptionWeight(hstring const& label, double value, double minimum, double maximum, double step)
        : m_label(label), m_value(value), m_minimum(minimum), m_maximum(maximum), m_step(step)
    {
    }

    hstring OptionWeight::Label() const
    {
        return m_label;
    }

    double OptionWeight::Value() const
    {
        return m_value;
    }

    void OptionWeight::Value(double value)
    {
        if (m_value == value) return;
        m_value = value;
        m_propertyChanged(*this, Microsoft::UI::Xaml::Data::PropertyChangedEventArgs{ L"Value" });
    }

    double OptionWeight::Minimum() const
    {
        return m_minimum;
    }

    double OptionWeight::Maximum() const
    {
        return m_maximum;
    }

    double OptionWeight::Step() const
    {
        return m_step;
    }

    winrt::event_token OptionWeight::PropertyChanged(
        Microsoft::UI::Xaml::Data::PropertyChangedEventHandler const& handler)
    {
        return m_propertyChanged.add(handler);
    }

    void OptionWeight::PropertyChanged(winrt::event_token const& token) noexcept
    {
        m_propertyChanged.remove(token);
    }
}

File: desktop/native-wpf/src/SurveyController.App/ViewModels/OptionWeight.cs
using System.ComponentModel;
using System.Runtime.CompilerServices;

namespace SurveyController.App.ViewModels;

/// <summary>权重行：Label/Minimum/Maximum/Step 不可变，Value 变更通知驱动比例预览刷新。</summary>
public sealed class OptionWeight : INotifyPropertyChanged
{
    private double _value;

    public OptionWeight(string label, double value, double minimum, double maximum, double step)
    {
        Label = label;
        _value = value;
        Minimum = minimum;
        Maximum = maximum;
        Step = step;
    }

    public string Label { get; }

    public double Minimum { get; }

    public double Maximum { get; }

    public double Step { get; }

    public double Value
    {
        get => _value;
        set
        {
            if (_value == value)
            {
                return;
            }
            _value = value;
            OnPropertyChanged();
        }
    }

    private string _percentageText = string.Empty;

    public string PercentageText
    {
        get => _percentageText;
        set
        {
            if (_percentageText == value)
            {
                return;
            }
            _percentageText = value;
            OnPropertyChanged();
        }
    }

    public event PropertyChangedEventHandler? PropertyChanged;

    private void OnPropertyChanged([CallerMemberName] string? propertyName = null) =>
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
}
